js/cards/cffa.ts

```diff
@@ -476,17 +473,20 @@ export default class CFFA implements Card, MassStorage<BlockFormat>, Restorable<
         return this.setBlockVolume(drive, disk);
     }
 
-    getBinary(drive: number): MassStorageData {
+    getBinary(drive: number): MassStorageData | null {
         drive = drive - 1;
+        if (!this._sectors[drive]?.length) {
+            return null;
+        }
         const blocks = this._sectors[drive];
         const data = new Uint8Array(blocks.length * 512);
         for (let idx = 0; idx < blocks.length; idx++) {
             data.set(blocks[idx], idx * 512);
         }
         return {
             name: this._name[drive],
-            ext: this._ext[drive],
-            data,
+            ext: 'po',
```

Comment:
@iflan Making this change because I remembered we don't deal with 2mg extensions on output at all yet, I started down the path of creating 2mg images but realized that's possibly overlap with what you're working on. If not, I can drop in some 2mg generating code pretty easily at least for the block devices.

Reply:
I'm fine with it either way. Once I have the tests in and factor out the WOZ and Nibble implementations so that they are easier to understand, then I'm going to start refactoring the loading and saving.

Here's a question for you, though. For saving a 2MG image, would you keep the original creator and so forth or would you change the creator to `a2js`? Likewise, in the situation where we're writing back to the same file, would you rewrite the whole file or _just_ the data section?

I'm been thinking about it like this:

There are two distinct functions:
* **Save** which saves a copy of the disk in the drive to "somewhere", like via the download dialog. This is a completely new copy of the disk and may even be a different format than the original. (For example, you could load a `.po` image and save it as a `.2mg` image.)
* **Persist** which constantly updates the original image when the disk is written. (There's no technical reason that it has to update the original image—it could just as easily update a copy—but this is how I've been thinking about it.)

When "saving", I think that an entirely new image should be created. So if the user saves over an existing file, it is essentially as if the file is deleted and then re-written.

On the other hand, in "persistent mode", the emulator should just update the minimum necessary to bring the image up to date. So, for example, if the image was a ProDOS-ordered block image, only the blocks that were written would actually be re-written.

What do you think? Do you have any strong opinions?

Reply:
I was pondering this, too, and I think I'm in agreement. I think most people would expect Persist behavior, and for the initial take on downloads I was thinking of preserving the creator field. But I feel it would be reasonable for something like a "Save As..." to use its own creator. Also, if we wanted to use the extended data field (almost nobody does, but I feel like somebody _should_), it would necessary to update the creator.